In [ ]:
#import dependencies

import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

Training Data

In [ ]:
train = pd.read_csv('/content/fraudTrain.csv')
print(train.head())

   Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   

      last gender                        street  ...      lat      long  \
0    Banks      F                561 Perry Cove  ...  36.0788  -81.1781   
1     Gill      F  43039 Riley Greens Suite 393  ...  48

In [ ]:
train.shape


(50601, 23)

In [ ]:
train.describe()

Unnamed: 0        cc_num           amt           zip           lat  \
count  50601.000000  5.060100e+04  50601.000000  50601.000000  50601.000000   
mean   25300.000000  4.194606e+17     71.639166  48624.144562     38.530921   
std    14607.394822  1.312479e+18    148.894633  27008.114706      5.096699   
min        0.000000  6.041621e+10      1.000000   1257.000000     20.027100   
25%    12650.000000  1.800429e+14      9.750000  25526.000000     34.668900   
50%    25300.000000  3.519607e+15     47.950000  48088.000000     39.371600   
75%    37950.000000  4.642255e+15     83.690000  72011.000000     41.846700   
max    50600.000000  4.992346e+18  11872.210000  99783.000000     65.689900   

               long      city_pop     unix_time     merch_lat    merch_long  \
count  50600.000000  5.060000e+04  5.060000e+04  50600.000000  50600.000000   
mean     -90.187505  8.944111e+04  1.326667e+09     38.532312    -90.190690   
std       13.919745  3.010694e+05  7.400696e+05      5.127440     13.934001   
min     -165.672300  2.300000e+01  1.325376e+09     19.040141   -166.670132   
25%      -96.790900  7.430000e+02  1.326007e+09     34.754728    -96.893008   
50%      -87.366700  2.456000e+03  1.326637e+09     39.360202    -87.330001   
75%      -80.065200  2.112500e+04  1.327265e+09     41.920227    -80.122978   
max      -67.950300  2.906700e+06  1.327923e+09     66.659242    -66.967742   

           is_fraud  
count  50600.000000  
mean       0.009111  
std        0.095015  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

Test Data

In [ ]:
test = pd.read_csv('/content/fraudTest.csv', on_bad_lines='skip')


In [ ]:
print(test.head())

   Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 

In [ ]:
test.shape

(120457, 23)

In [ ]:
test.describe()

Unnamed: 0        cc_num            amt            zip  \
count  120457.000000  1.204570e+05  120457.000000  120457.000000   
mean    60228.000000  4.170758e+17      69.402182   48828.538184   
std     34773.085025  1.308914e+18     154.395308   26842.555084   
min         0.000000  6.041621e+10       1.000000    1257.000000   
25%     30114.000000  1.800429e+14       9.630000   26292.000000   
50%     60228.000000  3.519607e+15      47.320000   48174.000000   
75%     90342.000000  4.634956e+15      82.920000   72042.000000   
max    120456.000000  4.992346e+18   13149.150000   99783.000000   

                 lat           long      city_pop     unix_time  \
count  120457.000000  120457.000000  1.204560e+05  1.204560e+05   
mean       38.536702     -90.214813  8.899253e+04  1.373605e+09   
std         5.069770      13.713548  3.021428e+05  1.055740e+06   
min        20.027100    -165.672300  2.300000e+01  1.371817e+09   
25%        34.668900     -96.798000  7.430000e+02  1.372641e+09   
50%        39.371600     -87.476900  2.408000e+03  1.373632e+09   
75%        41.940400     -80.175200  2.032800e+04  1.374474e+09   
max        65.689900     -67.950300  2.906700e+06  1.375463e+09   

           merch_lat     merch_long       is_fraud  
count  120456.000000  120456.000000  120456.000000  
mean       38.535821     -90.214889       0.003869  
std         5.105809      13.727944       0.062078  
min        19.042320    -166.671575       0.000000  
25%        34.749639     -96.892399       0.000000  
50%        39.373131     -87.431051       0.000000  
75%        41.959066     -80.267785       0.000000  
max        66.669356     -66.952352       1.000000

In [ ]:
# Checking for null values
train.isnull().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     1
city_pop                 1
job                      1
dob                      1
trans_num                1
unix_time                1
merch_lat                1
merch_long               1
is_fraud                 1
dtype: int64

In [ ]:
train.isna().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     1
city_pop                 1
job                      1
dob                      1
trans_num                1
unix_time                1
merch_lat                1
merch_long               1
is_fraud                 1
dtype: int64

In [ ]:
test.isnull().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 1
job                      1
dob                      1
trans_num                1
unix_time                1
merch_lat                1
merch_long               1
is_fraud                 1
dtype: int64

In [ ]:
test.isna().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 1
job                      1
dob                      1
trans_num                1
unix_time                1
merch_lat                1
merch_long               1
is_fraud                 1
dtype: int64

In [ ]:
train = train.dropna()
train.isnull().sum()


Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                0
merch_lat                0
merch_long               0
is_fraud                 0
dtype: int64

In [ ]:
train.isna().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                0
merch_lat                0
merch_long               0
is_fraud                 0
dtype: int64

In [ ]:
test = test.dropna()

In [ ]:
test.isnull().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                0
merch_lat                0
merch_long               0
is_fraud                 0
dtype: int64

In [ ]:
test.isna().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                0
merch_lat                0
merch_long               0
is_fraud                 0
dtype: int64

In [ ]:
X_train = train.drop(columns = ['is_fraud', 'trans_date_trans_time', 'cc_num', 'merchant', 'category', 'first', 'last', 'gender', 'street', 'city', 'state', 'job', 'dob', 'trans_num', 'unix_time', 'Unnamed: 0'])
Y_train = train['is_fraud']

In [ ]:
print(X_train)

          amt    zip      lat      long  city_pop  merch_lat  merch_long
0        4.97  28654  36.0788  -81.1781    3495.0  36.011293  -82.048315
1      107.23  99160  48.8878 -118.2105     149.0  49.159047 -118.186462
2      220.11  83252  42.1808 -112.2620    4154.0  43.150704 -112.154481
3       45.00  59632  46.2306 -112.1138    1939.0  47.034331 -112.561071
4       41.96  24433  38.4207  -79.4629      99.0  38.674999  -78.632459
...       ...    ...      ...       ...       ...        ...         ...
50595   74.07  33967  26.4722  -81.8122  224256.0  26.489912  -81.665144
50596   37.08  51632  40.7491  -95.0380    7297.0  40.539831  -94.772774
50597   26.74  20895  39.0298  -77.0793   19054.0  38.383734  -77.760825
50598    6.01  34112  26.1184  -81.7361  276002.0  25.144182  -81.658560
50599    3.11  76678  31.6591  -96.8094     263.0  31.283346  -97.152849

[50600 rows x 7 columns]


In [ ]:
print(Y_train)

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
50595    0.0
50596    0.0
50597    0.0
50598    0.0
50599    0.0
Name: is_fraud, Length: 50600, dtype: float64


In [ ]:
X_train.shape


(50600, 22)

In [ ]:
Y_train.shape

(50600,)

In [ ]:
X_test = test.drop(columns= ['Unnamed: 0','is_fraud', 'trans_date_trans_time', 'cc_num', 'merchant', 'category', 'first', 'last', 'gender', 'street', 'city', 'state', 'job', 'dob', 'trans_num', 'unix_time'])
Y_test = test['is_fraud']

In [ ]:
print(X_test)

          amt    zip      lat      long  city_pop  merch_lat  merch_long
0        2.86  29209  33.9659  -80.9355  333497.0  33.986391  -81.200714
1       29.84  84002  40.3207 -110.4360     302.0  39.450498 -109.960431
2       41.28  11710  40.6729  -73.5365   34496.0  40.495810  -74.196111
3       60.05  32780  28.5697  -80.8191   54767.0  28.812398  -80.883061
4        3.19  49632  44.2529  -85.0170    1126.0  44.959148  -85.884734
...       ...    ...      ...       ...       ...        ...         ...
120451  18.23  36869  32.4204  -85.0796   59744.0  32.922106  -84.947947
120452   8.44  77566  29.0393  -95.4401   28739.0  28.296802  -95.994634
120453  47.45   7208  40.6747  -74.2239  124967.0  40.384052  -73.564364
120454  71.64  45879  41.1410  -84.5722    6284.0  41.050883  -84.772206
120455   3.73  16329  41.8117  -79.2643     528.0  42.058187  -79.253352

[120456 rows x 7 columns]


In [ ]:
print(Y_test)

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
         ... 
120451    0.0
120452    0.0
120453    0.0
120454    0.0
120455    0.0
Name: is_fraud, Length: 120456, dtype: float64


In [ ]:
X_test.shape

(120456, 22)

In [ ]:
Y_test.shape

(120456,)

Scaling features

In [ ]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

Model fitting

Logistic Regression

In [ ]:
lr = LogisticRegression()

In [ ]:
lr.fit(X_train_scaled, Y_train)

LogisticRegression()

In [ ]:
y_pred_lr = lr.predict(X_test_scaled)
print(y_pred_lr)

[0. 0. 0. ... 0. 0. 0.]


In [ ]:
# Evaluate the model
print("Logistic Regression Accuracy:", accuracy_score(Y_test, y_pred_lr))
print(confusion_matrix(Y_test, y_pred_lr))
print(classification_report(Y_test, y_pred_lr))

Logistic Regression Accuracy: 0.9948196851962542
[[119831    159]
 [   465      1]]
              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00    119990
         1.0       0.01      0.00      0.00       466

    accuracy                           0.99    120456
   macro avg       0.50      0.50      0.50    120456
weighted avg       0.99      0.99      0.99    120456



Decision Tree

In [ ]:
decision_tree = DecisionTreeClassifier()

In [69]:
decision_tree.fit(X_train_scaled, Y_train)

DecisionTreeClassifier()

In [70]:
# Predict on the test set
y_pred_decision_tree = decision_tree.predict(X_test_scaled)
print(y_pred_decision_tree)

[0. 0. 0. ... 0. 0. 0.]


In [71]:
# Evaluate the model
print("Decision Tree Accuracy:", accuracy_score(Y_test, y_pred_decision_tree))
print(confusion_matrix(Y_test, y_pred_decision_tree))
print(classification_report(Y_test, y_pred_decision_tree))

Decision Tree Accuracy: 0.9908431294414558
[[119246    744]
 [   359    107]]
              precision    recall  f1-score   support

         0.0       1.00      0.99      1.00    119990
         1.0       0.13      0.23      0.16       466

    accuracy                           0.99    120456
   macro avg       0.56      0.61      0.58    120456
weighted avg       0.99      0.99      0.99    120456



 Random Forest

In [72]:
classifier =  RandomForestClassifier()

In [73]:
classifier.fit(X_train_scaled, Y_train)

RandomForestClassifier()

In [74]:
y_pred_randomforest = classifier.predict(X_test_scaled)
print(y_pred_randomforest)

[0. 0. 0. ... 0. 0. 0.]


In [75]:
# Evaluate the model
print("Random Forest Accuracy:", accuracy_score(Y_test, y_pred_randomforest))
print(confusion_matrix(Y_test, y_pred_randomforest))
print(classification_report(Y_test, y_pred_randomforest))

Random Forest Accuracy: 0.9955917513448894
[[119894     96]
 [   435     31]]
              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00    119990
         1.0       0.24      0.07      0.10       466

    accuracy                           1.00    120456
   macro avg       0.62      0.53      0.55    120456
weighted avg       0.99      1.00      0.99    120456



Logistic Regression Accuracy: **0.9948196851962542**


Decision Tree Accuracy: **0.9908431294414558**

Random Forest Accuracy: **0.9955917513448894**